src/Development/IDE/LSP/CodeAction.hs

```diff
@@ -134,6 +135,22 @@ suggestReplaceIdentifier contents Diagnostic{_range=_range@Range{..},..}
         = [ ("Replace with ‘" <> name <> "’", [mkRenameEdit contents _range name]) | name <- renameSuggestions ]
     | otherwise = []
 
+suggestNewDefinition :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
+suggestNewDefinition mb_contents Diagnostic{_message, _range}
+--     * Variable not in scope:
+--         suggestAcion :: Maybe T.Text -> Range -> Range
+    | Just contents <- mb_contents
+    , Just [name, typ] <- matchRegex (unifySpaces _message) "Variable not in scope: ([^ ]*) :: ([^*•]*)"
+    , sig <- name <> " :: " <> T.dropWhileEnd isSpace typ
+    , rangeLine <- _line (_start _range)
+    , contentLines <- T.lines contents
+    , nextEmptyLine <- findIndex T.null (drop rangeLine contentLines)
```

Comment:
Worth noting that this can fail. E.g. in
```
foo :: Bool -> String
foo True = bar "true"

foo False = "false"
```
it will insert the definition of `bar` before the second case of `foo`, which is illegal.

Do we have easy access to the name of the current definition (here `foo`)? If so, you could skip until the next top-level definition with a different name and insert before there.

Reply:
I don't think that the definition is immediately available, but we can extract it from the AST using the range. That should be easy enough, but perhaps I'll wait for #308 to merge to avoid conflicts.